File: frontend/src/types/content.ts
export type ContentStatus = "draft" | "approved" | "published";
export type ContentType = "post" | "article" | "campaign_brief" | "carousel";

export interface ContentItem {
  id: string;
  accountId: string;
  createdBy: string;
  title: string;
  body: string;
  contentType: ContentType;
  status: ContentStatus;
  imageUrl?: string | null;
  imageAssetRef?: string | null;
  sourceGenerationId?: string | null;
  createdAt: string;
  updatedAt: string;
  publishedAt?: string | null;
  version: number;
}
export interface GenerationRequest {
  prompt: string;
  model: "fast" | "quality";
  generateImage?: boolean;
  estimatedTokens?: number;
}
export interface GenerationChunk {
  type: "token" | "complete" | "failed" | "cancelled";
  value: string;
  generationId: string;
  imageUrl?: string | null;
}

export interface GenerationStart {
  jobId: string;
  channel: string;
  model: string;
  status: string;
  imageUrl?: string | null;
}

export interface GenerationJob {
  id: string;
  accountId: string;
  userId: string;
  model: string;
  status: string;
  prompt: string;
  response: string;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  costMicrousd: number;
  errorReason?: string | null;
  createdAt: string;
  startedAt?: string | null;
  completedAt?: string | null;
}

export interface PromptHistory {
  id: string;
  jobId: string;
  model: string;
  prompt: string;
  response: string;
  totalTokens: number;
  costMicrousd: number;
  createdAt: string;
}

export interface ImageGeneration {
  id: string;
  imageUrl: string;
  status: string;
}

File: content-service/src/content_service/schemas/content.py
from datetime import datetime
from typing import Literal
from uuid import UUID

import markdown
from pydantic import BaseModel, Field, computed_field

ContentStatus = Literal["draft", "approved", "published"]
ContentType = Literal["post", "article", "campaign_brief", "carousel"]

_MARKDOWN_EXTENSIONS = ["extra", "sane_lists", "nl2br"]


def render_markdown(body: str) -> str:
    return markdown.markdown(body, extensions=_MARKDOWN_EXTENSIONS)


class ContentCreate(BaseModel):
    title: str = Field(min_length=2, max_length=180)
    body: str = Field(min_length=1)
    content_type: ContentType = "post"
    image_url: str | None = None
    image_asset_ref: str | None = None


class ContentUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=2, max_length=180)
    body: str | None = Field(default=None, min_length=1)
    image_url: str | None = None
    image_asset_ref: str | None = None


class ContentRead(BaseModel):
    id: UUID
    account_id: UUID
    created_by: UUID
    title: str
    body: str
    content_type: str
    status: ContentStatus
    image_url: str | None
    image_asset_ref: str | None
    source_generation_id: UUID | None
    version: int
    created_at: datetime
    updated_at: datetime
    published_at: datetime | None

    model_config = {"from_attributes": True}

    @computed_field  # type: ignore[prop-decorator]
    @property
    def body_html(self) -> str:
        return render_markdown(self.body)


class ContentVersionRead(BaseModel):
    id: UUID
    content_id: UUID
    version: int
    title: str
    body: str
    status: str
    image_url: str | None
    image_asset_ref: str | None
    edited_by: UUID
    created_at: datetime

    model_config = {"from_attributes": True}

    @computed_field  # type: ignore[prop-decorator]
    @property
    def body_html(self) -> str:
        return render_markdown(self.body)


class ProductMetric(BaseModel):
    label: str
    value: str
    change: str


class ActivityRow(BaseModel):
    title: str
    detail: str
    status: str


class ProductView(BaseModel):
    title: str
    eyebrow: str
    description: str
    action: str
    metrics: list[ProductMetric]
    rows: list[ActivityRow | dict[str, str]]


class ContentCollection(BaseModel):
    view: ProductView
    items: list[ContentRead]
